### ISSUE TITLE
Support Memoizing Layers

### ISSUE THREAD

**Issue by adamgfraser** [zio/zio#2772]
We should implement a method to memoize a layer. This is important to allow a layer to be created once and provided to multiple services. This issue is currently impacting ZIO Test because we provide a separate copy of the test environment to each test and the test environment depends on the live environment, so each time we create a new test we create a new copy of the live environment, including allocating resources like a scheduler.

Thanks to @vasilmkd for identifying the issues through profiling of our internal test suites.

I will take the ticket.

**Comment by vasilmkd** [comment]
![zio-timers](https://user-images.githubusercontent.com/7115459/73545815-67e4dd00-443c-11ea-966c-a3da32353b2e.PNG)
![zio-memory-leak](https://user-images.githubusercontent.com/7115459/73545819-6a473700-443c-11ea-9e8d-c4764675d9b0.PNG)
![zio-heap-full](https://user-images.githubusercontent.com/7115459/73545822-6c10fa80-443c-11ea-9155-8e753ccaf202.PNG)

@adamgfraser Feel free to tag anyone here that might be interested in these.

**Comment by vasilmkd** [comment]
I'm also uploading the whole heap (4 Gb), I will update this comment when available.

**Comment by adamgfraser** [comment]
Following up on this, the `default` environments are generally `val` definitions so I don't think this problem is as bad as I feared, though this will still be an important feature for more advanced use cases with layers that do have meaningful costs to acquire and release and need to be shared among multiple dependencies.

**Comment by jdegoes** [comment]
I have some feedback on this. I will write it up.

**Comment by adamgfraser** [comment]
That would be great! I think we have tracked down the cause of the performance problems in our internal test suite to a particular test suite (will have a fix for that tomorrow) so we can focus on the right long term answer here.

**Comment by jdegoes** [comment]
@adamgfraser I too have been thinking about memoization. Not for performance, per se, but rather, to avoid construction of multiple implementations of the same resource.

Right now, if you use the same layer twice, e.g. `foo >>> bar`, `foo >>> baz`, then `foo` resources will be allocated twice. Now a simple `ZLayer#memoize` could let you solve that problem with, `foo.memoize.flatMap { foo => >... }`, but it feels less ergonomic.

Piggybacking on your idea of ZLayers as recipes, it should be possible to add memoize directly, such that `ZLayer#memoize` would return another `ZLayer`. I'll show how shortly...

**Comment by adamgfraser** [comment]
@jdegoes Love it!

**Comment by jdegoes** [comment]
This is just a sketch:

```scala
Managed[Nothing, MemoMap => ZManaged[RIn, E, ROut]]

trait MemoMap {
  def get[E, A, B](zlayer: ZLayer[A, E, B]): UIO[Option[B]]
  def memoize[E, A, B](zlayer: ZLayer[A, E, B], b: B): UIO[Unit]
}
```

The idea would be in the composition operators `>>>` and `++`, we check the memo map to see if the dependency is available. If it is, we use it. If it's not there, we obtain it, add a finalizer (to the outer `Managed`), and store it in the memo map. 

The `build` function would construct the empty memo map, backing it with a `Ref`, and then use that to build the whole thing.

**Comment by jdegoes** [comment]
^^^ @adamgfraser 

We could incorporate this via `ZLayer#shared`, so you could do something like `val sharedDb = Database.live.shared` to avoid allocating resources more than once for a database as it's fed into various other layers. 

Then atop this machinery for "sharing", we could add memoize, like `def memoize: UManaged[ZLayer[A, E, B]]`. This would do all the work upfront (lazily) and stuff the result into a "dumb" layer that would be much more efficient.

WIP.

**Comment by adamgfraser** [comment]
Closed by #2775.
